File: app/src/mock/java/com/beblue/gfpf/test/bebluegfpftest/data/FakeServiceApiImpl.java
package com.beblue.gfpf.test.bebluegfpftest.data;

import com.beblue.gfpf.test.bebluegfpftest.user.data.domain.GHSearchUser;
import com.beblue.gfpf.test.bebluegfpftest.user.data.domain.GHUser;
import com.beblue.gfpf.test.bebluegfpftest.user.data.service.ServiceApi;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;

import androidx.collection.ArrayMap;
import io.reactivex.Completable;
import io.reactivex.Single;

public class FakeServiceApiImpl implements ServiceApi {

    private static final ArrayMap<Integer, GHUser> GH_USER_SERVICE_DATA = FakeServiceApiClient.loadAllGHUsers();

    @Override
    public void createGHUser(GHUser... users) {
        for (GHUser GHUser : users) {
            GH_USER_SERVICE_DATA.put(GHUser.getId(), GHUser);
        }
    }

    @Override
    public Single<List<GHUser>> loadAllGHUsers() {
        /*try {
            Thread.sleep(2000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }*/

        //return Lists.newArrayList(GH_USER_SERVICE_DATA.values();

        Collection<GHUser> values = GH_USER_SERVICE_DATA.values();
        List<GHUser> list = new ArrayList<>(values);
        return Single.just(list);
    }

    @Override
    public Single<GHSearchUser> searchGHUserByName(String username, String sort, String order) {
        for (Map.Entry entry : GH_USER_SERVICE_DATA.entrySet()) {
            GHUser user = (GHUser) entry.getValue();

            List<GHUser> results = new ArrayList<>(1);
            results.add(user);

            GHSearchUser  searchUser = new GHSearchUser(1, false, user.getName(), results);
            searchUser.setUsers(results);

            String itemName = user.getLogin().trim().toLowerCase();
            username = username.trim().toLowerCase();

            if (itemName.equals(username) || itemName.contains(username)) {
                return Single.just(searchUser);
            }
        }

        return Single.error(new Throwable());
    }

    @Override
    public Single<GHUser> loadGHUserById(int id) {
        for (Map.Entry entry : GH_USER_SERVICE_DATA.entrySet()) {
            GHUser user = (GHUser) entry.getValue();

            int itemId = user.getId();

            if (itemId == id) {
                return Single.just(user);
            }
        }

        return Single.error(new Throwable());
    }

    @Override
    public Completable updateGHUser(int id, String GHUser) {
        return null;
    }

    @Override
    public Completable deleteGHUser(int id) {
        return null;
    }
}

File: github_api/src/mock/java/com.gfpf.github_api/data/FakeServiceApiImpl.kt
package com.gfpf.github_api.data

import androidx.collection.ArrayMap
import com.gfpf.github_api.data.service.GHServiceApi
import com.gfpf.github_api.domain.user.GHRepository
import com.gfpf.github_api.domain.user.GHSearchUser
import com.gfpf.github_api.domain.user.GHTag
import com.gfpf.github_api.domain.user.GHUser
import java.util.Locale

class FakeServiceApiImpl : GHServiceApi {
    override suspend fun createGHUser(vararg users: GHUser) {
        for (user in users) {
            GH_USER_SERVICE_DATA[user.id] = user
        }
    }

    override suspend fun loadAllGHUsers(): List<GHUser> {
        return ArrayList(GH_USER_SERVICE_DATA.values)
    }

    override suspend fun loadGHUserById(id: Int): GHUser? {
        return GH_USER_SERVICE_DATA[id]
    }

    override suspend fun searchGHUserByName(
        username: String,
        sort: String,
        order: String
    ): GHSearchUser {
        val results = GH_USER_SERVICE_DATA.values.filter { user ->
            val itemName = user.login?.trim()?.lowercase(Locale.getDefault())
            val searchTerm = username.trim().lowercase(Locale.getDefault())
            itemName == searchTerm || itemName?.contains(searchTerm) == true
        }

        return GHSearchUser(results.size, false, results)
    }

    override suspend fun loadGHUserRepos(id: String): List<GHRepository> {
        val userId = id.toIntOrNull()
        return GH_REPOS_SERVICE_DATA[userId] ?: emptyList()
    }

    override suspend fun loadGHRepositoryTags(owner: String, repo: String): List<GHTag> {
        val repoKey = "$owner/$repo"
        return GH_TAGS_SERVICE_DATA[repoKey] ?: emptyList()
    }

    override suspend fun updateGHUser(id: Int, ghUser: String) {
    }

    override suspend fun deleteGHUser(id: Int) {
        GH_USER_SERVICE_DATA.remove(id)
    }

    companion object {
        private val GH_USER_SERVICE_DATA: ArrayMap<Int?, GHUser> =
            FakeServiceApiClient.loadAllGHUsers()

        private val GH_REPOS_SERVICE_DATA: ArrayMap<Int?, List<GHRepository>> =
            FakeServiceApiClient.loadAllGHRepositories()

        private val GH_TAGS_SERVICE_DATA: Map<String, MutableList<GHTag>> =
            FakeServiceApiClient.loadAllGHTags().mapKeys { (userId, _) -> userId.toString() }

        /*private val GH_TAGS_SERVICE_DATA: ArrayMap<Int?, List<GHTag>> =
            FakeServiceApiClient.loadAllGHTags()*/
    }
}